lib/protobuf/field/message_field.rb

```diff
@@ -35,8 +35,8 @@ def coerce!(value)
           nil
         elsif value.is_a?(type_class)
           value
-        elsif value.respond_to?(:to_proto)
-          value.to_proto
+        elsif value.respond_to?(:to_proto) && (proto_value = value.to_proto).is_a?(type_class)
```

Comment:
I'm actually good with the proposal, it might be more readable and sounds like it's faster? I changed it as suggested.